locations/middlewares/anti_bot_detection.py

```diff
@@ -0,0 +1,103 @@
+    # https://developers.cloudflare.com/bots/
+    CLOUDFLARE = {"name": "Cloudflare", "zyte_bypassable": True}
+
+    # DataDome bot protection product brochure:
+    # https://datadome.co/products/bot-protection/
+    DATADOME = {"name": "DataDome", "zyte_bypassable": True}
+
+    # HUMAN Bot Defender bot protection product brochure:
+    # https://www.humansecurity.com/products/human-bot-defender
+    HUMAN = {"name": "HUMAN", "zyte_bypassable": True}
+
+    # Imperva bot protection product brochure:
+    # https://www.imperva.com/products/bot-detection-mitigation/
+    IMPERVA = {"name": "Imperva", "zyte_bypassable": True}
+
+    # Qrator bot protection documentation:
+    # https://docs.qrator.net/en/technologies/tracking-cookie.html
+    QRATOR = {"name": "Qrator", "zyte_bypassable": True}
+
+
+class AntiBotDetectionMiddleware:
+    def __init__(self, crawler):
+        self.crawler = crawler
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def process_response(self, request: Request, response: Response, spider: Spider):
+        if response.headers.get("cf-mitigated"):
+            self.add_anti_bot_method(AntiBotMethods.CLOUDFLARE, spider)
+
+        if response.headers.get("x-datadome"):
+            self.add_anti_bot_method(AntiBotMethods.DATADOME, spider)
+
+        if server := response.headers.get("server"):
+            server_utf8 = self.decode_http_header_value(server)
+            if server_utf8.upper() == "QRATOR":
+                self.add_anti_bot_method(AntiBotMethods.QRATOR, spider)
+            if response.status == 403 and server_utf8.upper() == "MICROSOFT-AZURE-APPLICATION-GATEWAY/V2":
+                self.add_anti_bot_method(AntiBotMethods.AZURE_WAF, spider)
+
+        if cookies := response.headers.getlist("set-cookie"):
+            for cookie in cookies:
+                cookie_utf8 = self.decode_http_header_value(cookie)
+                if cookie_utf8.startswith("visid_incap_") or cookie_utf8.startswith("incap_ses_"):
+                    self.add_anti_bot_method(AntiBotMethods.IMPERVA, spider)
+
+        if response.status == 403:
+            # We may need to decompress the response body before it
+            # can be parsed to detect anti-bot methods. This is
+            # necessary because HttpErrorMiddleware is ordered
+            # before HttpCompressionMiddleware per:
+            # 1. https://docs.scrapy.org/en/latest/topics/settings.html#spider-middlewares-base
+            # 2. https://docs.scrapy.org/en/latest/topics/settings.html#downloader-middlewares-base
+            # It is probably not a good idea to reorder these
+            # default middleware orders.
+            httpcm = HttpCompressionMiddleware.from_crawler(spider.crawler)
+            decompressed_response = httpcm.process_response(request, response, spider)
+            html_response = HtmlResponse(decompressed_response.url, body=decompressed_response.body)
```

Comment:
Not important because html_response goes nowhere, but response.replace would be  better if it was used

Reply:
This is a good suggestion because some spiders may (unlikely though) gracefully handle HTTP 403 errors, for example, to pause and queue a retry to avoid rate limiting mechanisms. In such situation, we'd be decompressing the body of the response two times.

I'd have to look into this deeper because I suspect we may need to replace not just the body (compressed -> uncompressed data) but also modify the headers to omit any mention of the body being compressed.